src/ssl.c

```diff
@@ -20796,6 +20828,160 @@ int wolfSSL_RSA_GenAdd(WOLFSSL_RSA* rsa)
 }
 #endif /* NO_RSA */
 
+#ifdef WOLFSSL_SIGNAL
+int wolfSSL_HMAC_CTX_init(HMAC_CTX* ctx)
+{
+    WOLFSSL_MSG("wolfSSL_HMAC_CTX_init");
+    (void) ctx;
+
+    return SSL_SUCCESS;
+}
+
+
+int wolfSSL_HMAC_Init(WOLFSSL_HMAC_CTX* ctx, const void* key, int keylen,
+                  const EVP_MD* type)
+{
+    int hmac_error = 0;
+
+    WOLFSSL_MSG("wolfSSL_HMAC_Init");
+
+    if (ctx == NULL) {
+        WOLFSSL_MSG("no ctx on init");
+        return SSL_FAILURE;
+    }
+
+    if (type) {
+        WOLFSSL_MSG("init has type");
+
+        if (XSTRNCMP(type, "MD5", 3) == 0) {
+            WOLFSSL_MSG("md5 hmac");
+            ctx->type = MD5;
+        }
+        else if (XSTRNCMP(type, "SHA256", 6) == 0) {
+            WOLFSSL_MSG("sha256 hmac");
+            ctx->type = SHA256;
+        }
+
+        /* has to be last since would pick or 256, 384, or 512 too */
+        else if (XSTRNCMP(type, "SHA", 3) == 0) {
+            WOLFSSL_MSG("sha hmac");
+            ctx->type = SHA;
+        }
+        else {
+            WOLFSSL_MSG("bad init type");
+            return SSL_FAILURE;
+        }
+    }
+
+    if (key && keylen) {
+        WOLFSSL_MSG("keying hmac");
+
+        if (wc_HmacInit(&ctx->hmac, NULL, INVALID_DEVID) == 0) {
+            hmac_error = wc_HmacSetKey(&ctx->hmac, ctx->type, (const byte*)key,
+                                       (word32)keylen);
+            if (hmac_error < 0)
```

Comment:
If wc_HmacSetKey() fails here, do we also need to call wc_HmacFree() to free the "&ctx->hmac" initialized above?  Or are we expecting the calling application to call wolfSSL_HMAC_cleanup() in the case where wolfSSL_HMAC_Init() fails?

Reply:
Thank you for your suggestion.
I added wc_HMAC_Free call when wc_HmacSetKey failed.
Yes, because wc_HmacInit is called inside wolfSSL_HMAC_Init, it should be freed in this function.
In old OpenSSL API, it does not return error code, so you didn't worry about error routine.
Like Signal, recent OpenSSL applications adopt new OpenSSL library.
I guess this kind of issue will happen in wolfSSL OpenSSL Compatibility API from now on.